### PR TITLE
[BLCKCHN-280] Refactor for Unit Testing for Developer Platform SDK (PY) #17

### PR DESCRIPTION
**Pull request by leejw51crypto**
tidy up

# Pull Request Template

## Description
Please include a summary of the change and which issue is fixed. Include the context and motivations behind your changes.

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:
Before you submit the pull request, please review the following checklist:
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

## Linked Issues
Mention the issues that are addressed by this PR. You can use the `#` symbol to link an issue.

### CODE REVIEW

**Comment by leejw51crypto** [comment]
testing github action

**Comment by leejw51crypto** [comment]
polishing code

**Review by yeehan-crypto-com — APPROVED** [review]
Great job - 92% test coverage, maybe can consider increasing test coverage threshold from 62?

File: crypto_com_developer_platform_client/client.py [hunk 1]
```diff
@@ -6,6 +6,13 @@ class Client:
     _api_key: str
     _provider: str
 
+    def __init__(self):
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/client.py]: why do we need __init__ with pass?

File: crypto_com_developer_platform_client/client.py [hunk 2]
```diff
@@ -41,26 +48,60 @@ def init(cls, api_key: str, provider: str = "") -> None:
         CronosId.init(cls())
 
     @classmethod
-    def get_api_key(cls) -> str:
+    def get_api_key_class(cls) -> str:
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/client.py]: Not sure why change the name? Is it consistent with the TS naming convention?

File: crypto_com_developer_platform_client/client.py [hunk 3]
```diff
@@ -41,26 +48,60 @@ def init(cls, api_key: str, provider: str = "") -> None:
...
         """
-        if cls._api_key is None:
+        if not hasattr(cls, "_api_key") or cls._api_key is None:
             raise ValueError("API key is not set. Please set the API key.")
 
         return cls._api_key
 
     @classmethod
-    def get_provider(cls) -> str:
+    def get_provider_class(cls) -> str:
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/client.py]: Not sure why we change the name here

File: crypto_com_developer_platform_client/client.py [hunk 4]
```diff
@@ -41,26 +48,60 @@ def init(cls, api_key: str, provider: str = "") -> None:
...
+
+    def get_provider(self) -> str:
+        """
+        Get the provider from instance (delegates to class method).
+
+        :return: The provider.
+        """
+        return self.__class__.get_provider_class()
+
+    def get_api_key_instance(self) -> str:
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/client.py]: same as get_api_key

File: crypto_com_developer_platform_client/exchange.py [hunk 5]
```diff
@@ -26,7 +26,10 @@ def get_all_tickers(cls) -> ApiResponse:
 
         :return: A list of all available tickers and their information.
         """
-        return get_all_tickers()
+        if cls._client is None:
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/exchange.py:29]: this is not needed if if we check the condition directly in the client file

File: crypto_com_developer_platform_client/client.py (hunk 1 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/client.py]: it's for instance, restoring back to keep compatible

File: crypto_com_developer_platform_client/client.py (hunk 2 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/client.py]: changing back

File: crypto_com_developer_platform_client/client.py (hunk 3 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/client.py]: changing back

File: crypto_com_developer_platform_client/integrations/cronosid_api.py [hunk 6]
```diff
@@ -22,10 +23,13 @@ def resolve_cronos_id(api_key: str, name: str) -> ApiResponse:
     )
 
     if response.status_code not in (200, 201):
-        error_body = response.json()
-        server_error_message = (
-            error_body.get("error") or f"HTTP error! status: {response.status_code}"
-        )
+        try:
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/integrations/cronosid_api.py]: The endpoint returns status: success, data or status: failed, message. If we want to add Try Catch it would only be for scenarios where the response does not return a json response. 

I am not opposing Try Catch, but if we decide to make this change it should be in a separate ticket and in parallel we should update the TS library so they stay insync and we avoid technical dept down the road.

**Review by rarcifa — COMMENTED** [review]
Thanks @leejw51crypto — this is a really in-depth and well-structured unit test setup. My only concern is that this library is public and client-facing. Since we're already doing both unit and integration testing on the backend to fully validate all endpoints, I think duplicating tests on the client side has some drawbacks:

It increases the code size and technical debt we have to maintain.

It adds additional dependencies that external developers may not want or need.

Because we already handle all the testing on the server side, the client just acts as a lightweight interface. My opinion is that we should keep the public client as minimal and dependency-free as possible — and trust our backend test coverage to ensure correctness.

Let me know your thoughts, but I’d prefer to keep testing scoped to the backend for now.

**Comment by leejw51crypto** [comment]
will move testing code to developer platform service.
this pr will only contain other changes

File: crypto_com_developer_platform_client/client.py (hunk 4 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/client.py]: get_api_key is class method, this one is instance method.

File: crypto_com_developer_platform_client/exchange.py (hunk 5 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/exchange.py:29]: before calling get_api_instance(), it check validity of _client

File: crypto_com_developer_platform_client/client.py (hunk 4 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/client.py]: changing back

File: crypto_com_developer_platform_client/exchange.py (hunk 5 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/exchange.py:29]: get_all_tickers need apikey, so this change ensures, apikey is given correctly

File: crypto_com_developer_platform_client/integrations/cronosid_api.py (hunk 6 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/integrations/cronosid_api.py]: try catch is necessary to write efficient unit test

Comment (leejw51crypto) [reply crypto_com_developer_platform_client/integrations/cronosid_api.py]:
```
        error_body = response.json()
        server_error_message = (
            error_body.get("error") or f"HTTP error! status: {response.status_code}"
        )
```
for example, by some reason, json format is broken, it throw exception 
new code handle it, process it to return error message

Comment (leejw51crypto) [reply crypto_com_developer_platform_client/integrations/cronosid_api.py]: and unit test process , in edge cases easily

**Comment by leejw51crypto** [comment]
unit test result

Required test coverage of 80% reached. Total coverage: 91.38%

**Comment by leejw51crypto** [comment]
porting test code to https://github.com/crypto-com/developer-platform-sdk-service

File: crypto_com_developer_platform_client/integrations/wallet_api.py [hunk 7]
```diff
@@ -22,10 +22,14 @@ def create_wallet(api_key: str) -> ApiResponse:
...
+            error_body = response.json()
+            server_error_message = (
+                error_body.get("error") or f"HTTP error! status: {response.status_code}"
+            )
+        except ValueError:
+            # Handle non-JSON error responses
+            server_error_message = f"HTTP error! status: {response.status_code}"
         raise Exception(server_error_message)
 
     return response.json()
```
Comment (eric-cw-crypto) [inline crypto_com_developer_platform_client/integrations/wallet_api.py]: looks like these code are duplicating. do you think is better to put into a shared function instead?

**Review by eric-cw-crypto — APPROVED** [review]

File: crypto_com_developer_platform_client/integrations/wallet_api.py (hunk 7 above)
Comment (leejw51crypto) [reply crypto_com_developer_platform_client/integrations/wallet_api.py]: removed redundant code
also catch mal-formed json error

File: crypto_com_developer_platform_client/exchange.py (hunk 5 above)
Comment (rarcifa) [reply crypto_com_developer_platform_client/exchange.py:29]: You are right this is needed after our latest change and is consistent with the other methods.

File: crypto_com_developer_platform_client/integrations/api_interfaces.py [hunk 8]
```diff
@@ -1,12 +1,33 @@
...
+    SUCCESS = "Success"
+    FAILED = "Failed"
 
 
 class ApiResponse(TypedDict):
     status: Status
     data: Dict[str, Any]
+
+
+def handle_api_error(response: requests.Response) -> None:
```
Comment (rarcifa) [inline crypto_com_developer_platform_client/integrations/api_interfaces.py]: I do agree with reducing duplication as much as possible, but there are cases where it is ok to repeat logic. Maybe we can move this into a helper function instead of interface.

**Review by rarcifa — APPROVED** [review]

**Comment by leejw51crypto** [comment]
modifying for feedback
